khard/khard.py

```diff
@@ -296,47 +296,118 @@ def list_address_books(address_books):
+        if not parsable:
+            print("Address book: {}".format(selected_address_books[0]))
+        table_header = ["index", "name", "phone", "email"]
     else:
-        print("Address books: {}".format(', '.join(
-            [str(book) for book in selected_address_books])))
-        table_header = ["Index", "Name", "Phone", "E-Mail", "Address book"]
+        if not parsable:
+            print("Address books: {}".format(', '.join(
+                [str(book) for book in selected_address_books])))
+        table_header = ["index", "name", "phone", "email", "addressbook"]
     if config.show_uids:
-        table_header.append("UID")
-        abook_collection = AddressBookCollection('short uids collection',
-                                                 selected_address_books)
+        table_header.append("uid")
+
+    if parsable:
+        # Legacy default header fields for parsable.
+        table_header = ["uid", "name", "addressbook"]
+
+    if fields:
+        table_header = [ugly_name(x) for x in fields]
+
+    abook_collection = AddressBookCollection('short uids collection',
+                                             selected_address_books)
 
-    table.append(table_header)
+    if not parsable:
+        table.append([pretty_name(x) for x in table_header])
     # table body
     for index, vcard in enumerate(vcard_list):
         row = []
-        row.append(index + 1)
+        for field in table_header:
+            if field == 'index':
+                row.append(index + 1)
+            elif field in ['name', 'phone', 'email']:
+                row.append(get_special_field(vcard, field))
+            elif field == 'addressbook':
+                row.append(vcard.address_book.name)
+            elif field == 'uid':
+                if parsable:
+                    row.append(vcard.uid)
+                elif abook_collection.get_short_uid(vcard.uid):
+                    row.append(abook_collection.get_short_uid(vcard.uid))
+                else:
+                    row.append("")
+            else:
+                attr_name = field.split('.')[0]
+                val = ''
+                if hasattr(vcard, attr_name):
+                    val = getattr(vcard, attr_name)
+                    # Nest multiple keys like emails.home.1
+                    for partial in field.split('.')[1:]:
+                        if isinstance(val, dict) and partial in val:
+                            val = val[partial]
+                        elif partial.isdigit() and isinstance(val, list) \
+                                and len(val) > int(partial):
+                            val = val[int(partial)]
+                        # TODO: Completely support case insensitive indexing
+                        elif isinstance(val, dict) and partial.upper() in val:
+                            val = val[partial.upper()]
+                        else:
+                            val = ''
+                row.append(val)
```

Comment:
Maybe you could put this in a separate function `get_nested_field`.  That would make it easier to write unit tests for this as it seems like an important piece of new code. What do you think?

Reply:
Done. In what file should the tests be written for this function?

Reply:
I think you could start a new test class in `test/test_command_line_interface.py` which is named after the function and has tests for some important cases of its behaviour.

Reply:
Sounds like a good idea.

This merge request is getting pretty big though. I don't know when/if I'll get around to writing comprehensive tests for that function.